libminifi/test/flow-tests/CustomProcessors.h

```diff
@@ -0,0 +1,106 @@
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef NIFI_MINIFI_CPP_CUSTOMPROCESSORS_H
+#define NIFI_MINIFI_CPP_CUSTOMPROCESSORS_H
+
+#include <unordered_map>
+#include <string>
+#include <random>
+#include <YamlConfiguration.h>
+#include "core/Processor.h"
+#include "TestBase.h"
+#include "../../extensions/standard-processors/processors/GenerateFlowFile.h"
+
+namespace org {
+namespace apache {
+namespace nifi {
+namespace minifi {
+namespace processors {
+
+static core::Relationship Apple{"apple", ""};
+static core::Relationship Banana{"banana", ""};
+// The probability that this processor routes to Apple
+static core::Property AppleProbability = core::PropertyBuilder::createProperty("AppleProbability")->withDefaultValue<int>(100)->build();
+// The probability that this processor routes to Banana
+static core::Property BananaProbability = core::PropertyBuilder::createProperty("BananaProbability")->withDefaultValue<int>(0)->build();
+
+class ProcessorWithStatistics {
+ public:
+  std::atomic<int> trigger_count{0};
+};
+
+class TestProcessor : public core::Processor, public ProcessorWithStatistics {
+ public:
+  TestProcessor(std::string name, utils::Identifier &uuid) : Processor(name, uuid) {}
+  TestProcessor(std::string name) : Processor(name) {}
+  void initialize() override {
+    setSupportedProperties({AppleProbability, BananaProbability});
+    setSupportedRelationships({Apple, Banana});
+  }
+  void onTrigger(const std::shared_ptr<core::ProcessContext> &context, const std::shared_ptr<core::ProcessSession> &session) override {
+    ++trigger_count;
+    auto flowFile = session->get();
+    if (!flowFile) return;
+    std::random_device rd{};
+    std::uniform_int_distribution<int> dis(0, 100);
+    int rand = dis(rd);
+    if (rand <= apple_probability_) {
+      session->transfer(flowFile, Apple);
+      return;
+    }
+    rand -= apple_probability_;
+    if (rand <= banana_probability_) {
+      session->transfer(flowFile, Banana);
+      return;
+    }
+    throw std::runtime_error("Couldn't route file");
+  }
+  void onSchedule(const std::shared_ptr<core::ProcessContext> &context, const std::shared_ptr<core::ProcessSessionFactory> &sessionFactory) override {
+    int apple;
+    assert(context->getProperty(AppleProbability.getName(), apple));
+    int banana;
+    assert(context->getProperty(BananaProbability.getName(), banana));
```

Comment:
Without nodebug ifdef I'm not sure these assert actually ever do something.

Reply:
done